Add error page for initial REST request failure & Abort further load

florian-h05 opened the pull request:
Migitates and therefore closes #1205.

ghys commented:
Thanks - not a fan of `alert()` though.
For HABot I used the Quasar framework had a special page provided that would be redirected to when these situations occurred (they would be accessible even when added as a PWA still, they could be a routable page or a pure static HTML page):

![image](https://github.com/openhab/openhab-webui/assets/2004147/d2b62721-b7f6-4c4a-b211-5cddddf9cae6)

florian-h05 commented:
Good idea, I will try to have a look at this.

ghys reviewed (commented):
Great! Let me know if you think it's better to merge this still, it's an improvement over a dialog that didn't show but `alert()` is IMO too obtrusive.

florian-h05 commented:
@ghys I've added such a "page" (it is no real page, but instead part of app.vue, but anyway it should work fine):

![image](https://github.com/openhab/openhab-webui/assets/73423173/f5b95305-95d0-43d6-997e-f62ae82c6dff)

florian-h05 commented:
I have tested this on my dev system by deploying the UI artifact and then stopping the `org.openhab.core.io.rest.core` bundle, works fine!

ghys commented on bundles/org.openhab.ui/web/src/components/app.vue:
```diff
@@ -684,6 +710,35 @@ export default {
       return toast
+    },
+    purgeServiceWorkerAndCaches () {
+      navigator.serviceWorker.getRegistrations().then(function (registrations) {
+        for (let registration of registrations) {
+          registration.unregister().then(function () {
+            return self.clients.matchAll()
+          }).then(function (clients) {
+            clients.forEach(client => {
+              if (client.url && 'navigate' in client) {
+                setTimeout(() => { client.navigate(client.url.split('#')[0]) }, 1000)
+              }
+            })
+          })
+        }
+      })
+      window.caches.keys().then(function (cachesNames) {
```
In about.vue there's some logic to only show the button when it's relevant:

https://github.com/openhab/openhab-webui/blob/013b23fc70c367900c1b874b16fb7df3f3b462e5/bundles/org.openhab.ui/web/src/pages/about.vue#L162-L175

I thought it was better not to show it if you don't have a service worker registered or cached assets which are only available if you access the UI on a secure origin; otherwise I think it will just confuse the user as the button will do nothing (and incidentally this code will break because `navigator.serviceWorker` and `window.caches` won't be defined). This is the most probable situation when users access OH on HTTP port 8080 in a local network and this option makes no sense in this situation.

Even if you access OH on HTTPS with port 8443 it still wouldn't work if the certificate isn't trusted. The other option is to force your OH URL as a secure location manually with e.g. [chrome://flags/#unsafely-treat-insecure-origin-as-secure](chrome://flags/#unsafely-treat-insecure-origin-as-secure), which is obviously not formally recommended.

ghys reviewed (commented):
Thanks, I like the different options for reloading at different levels!
In your last commit you reordered the buttons in a more logical manner (in increasing order of "strength") but also removed the warning on purging the cache, can you elaborate why? I thought it was rather nice to warn the user. Can you elaborate why?

florian-h05 commented:
> also removed the warning on purging the cache, can you elaborate why? I thought it was rather nice to warn the user.

Yeah, but when deploying the UI bundle to my dev server (instead of running the dev server) I noticed that the warning was not shown. I guess it is a CSS visibility problem, I will have to check that when I’m back home.

florian-h05 commented:
@ghys I have added the logic to only show the "Purge Caches And Reload" button from about.vue and re-introduced the confirmation dialog. It was not visible because of CSS visibility set on for f7-app, I fixed that and also cleaned up app.vue a bit. From my side, this is ready for merging now.

florian-h05 commented:
@ghys I've fixed the conflict, can you please have a look?

ghys commented:
Uhh.. there are conflicts again. Probably due to me merging #2075. Could you please have a look again (sorry)?

I also think we should start to offload some methods out of `app.vue` to other files, it's becoming quite bulky - I mentioned it on another PR.

florian-h05 commented:
Yeah, I've already read your comment on that PR. I will have a look at refactoring that error handling logic to a mixin, I will ping you when I am finished.

florian-h05 commented:
@ghys I have refactored this now: The reload stuff moved to the reload-mixin, which is now also used by about.vue and the connection health stuff from app.vue moved to the connection-health-mixin.

As I have done that refactoring from my iPad (I am currently on a short trip to a concert) I haven‘t been able to test this, so please test:

- About page: „Reload“ and „Purge Service Worker and Caches“
- app.vue: Whether the error page still completely works, and whether the SSE communication failure toast still works

ghys commented on bundles/org.openhab.ui/web/src/components/connection-health-mixin.js:
```diff
@@ -0,0 +1,94 @@
+import mixin from 'reload-mixin.js'
```
Thanks - this could have waited, enjoy your concert! (who is it?)
I'll tentatively do the testing tomorrow. The builds still fail, though:

```
[INFO] ERROR in ./src/components/connection-health-mixin.js
[INFO] Module not found: Error: Can't resolve 'reload-mixin.js' in '/home/runner/work/openhab-webui/openhab-webui/bundles/org.openhab.ui/web/src/components'
[INFO]  @ ./src/components/connection-health-mixin.js 1:0-36 3:11-16
```


```suggestion
import mixin from './reload-mixin'
```
As an habit I dropped the .js extensions in imports, except in `assets` where is it required to build the component docs (as it is run by Node which requires them).

florian-h05 commented:
> Thanks - this could have waited, enjoy your concert! (who is it?)

We had to drive 6 hours though Germany, and I wasn't driving so I had plenty of free time.
[Two Steps From Hell](https://www.twostepsfromhell.com/) really performed an epic concert, absolutely great. I can really recommend checking out their music!

> I'll tentatively do the testing tomorrow. The builds still fail, though:

I have just tested everything I mentioned above, fixed the build and addressed your review regading file extensions.

ghys commented:
> [Two Steps From Hell](https://www.twostepsfromhell.com/) really performed an epic concert, absolutely great. I can really recommend checking out their music!

This was what appeared on Spotify when I checked them out, so I guess it was already my kind of stuff :) I'm sure it was indeed epic.

![image](https://github.com/openhab/openhab-webui/assets/2004147/90fa64fa-de96-4f3e-bde8-ab0cc8a72c87)

I think we're good here!
